spec/consensus/proposer-based-timestamp/pbts-algorithm_002_draft.md

```diff
@@ -0,0 +1,138 @@
+the clock of `p` is `PRECISION` *ahead* of the clock of the proposer of `v`, and the proposal's transmission delay is `MSGDELAY` (maximum).
+
+## Updated Consensus Algorithm
+
+The following changes are proposed for the algorithm in the [arXiv paper][arXiv].
+
+#### New `StartRound`
+
+There are two additions to the `propose` round step when executed by the `proposer` of a round:
+
+1. to ensure time monotonicity, the proposer does not propose a value until its current local time becomes greater than the previously decided value's time
+1. when the proposer produce a new proposal it sets the proposal's time to its current local time
+   - no changes are made to the logic when a proposer has a non-nil `validValue`, which retains its original proposal time.
+
+#### **[PBTS-ALG-STARTROUND.1]**
+
+```go
+function StartRound(round) {
+ round_p ← round
+ step_p ← propose
+ if proposer(h_p, round_p) = p {
+  wait until now_p > decision_p[h_p-1].time // time monotonicity
+  if validValue_p != nil {
+   proposal ← validValue_p
+  } else {
+   proposal ← getValue()
+   proposal.time ← now_p // proposal time
+  }
+   broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+ } else {
+  schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+ }
+}
+```
+
+#### New Rule Replacing Lines 22 - 27
+
+The rule on line 22 applies to values `v` proposed for the first time, i.e., for proposals not backed by `2f + 1 PREVOTE`s for `v` in a previous round.
+The `PROPOSAL` message, in this case, carry `-1` in its `validRound` field.
+
+The new rule for issuing a `PREVOTE` for a proposed value `v` requires the value to be `timely`.
+As the `timely` predicate is evaluated in the moment that the value is received,
+as part of a `PROPOSAL` message, we require the `PROPOSAL` message to be `timely`.
+
+#### **[PBTS-ALG-UPON-PROP.1]**
+
+```go
+upon timely(⟨PROPOSAL, h_p, round_p, v, −1⟩) from proposer(h_p, round_p) while step_p = propose do {
+  if valid(v) ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+    broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩ 
+  }
+  else {
+    broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+  }
+  step_p ← prevote
+}
+```
+
+#### Rules at Lines 28 - 33 remain unchanged
+
+The rule on line 28 applies to values `v` proposed again in the current round because its proposer received `2f + 1 PREVOTE`s for `v` in a previous round `vr`.
+This means that there was a round `r <= vr` in which `2f + 1` processes accepted `v` for the first time, and so sent `PREVOTE`s for `v`.
+Which, in turn, means that these processes executed the line 22 of the algorithm, and therefore judged `v` as a `timely` proposal.
+
+In other words, we don't need to verify whether `v` is a timely proposal because at least `f + 1` processes judged `v` as `timely` in a previous round,
+and because, since `v` was re-proposed as a `validValue` (line 16), `v.time` has not being updated from its original proposal.
+
+**All other rules remains unchanged.**
```

Comment:
If I'm understanding correctly, this means that we should evaluate timely on line 36, even when validRound is -1. I think this is fine, since we're seeing the +2/3 prevotes, but it could get mention in this design.

Reply:
I think you meant "we should **not** evaluate timely on line 36", right?

Reply:
It is a good point. I didn't mention lines 36 (Prevote -> Precommit) and 49 (Precommit -> Commit) because I think we don't need any additional logic there. But indeed we are assuming that `untimely` proposals can be accepted, provided that justified by a proper quorum of Prevotes.

Reply:
sorry for the typo, thank you for the correction @josef-widder

Reply:
> But indeed we are assuming that untimely proposals can be accepted, provided that justified by a proper quorum of Prevotes.

I am not sure what you mean. I guess if there is an untimely proposal that is justified by a proper quorum of prevotes that means we are outside the n>3f assumption, no? (or outside the synchrony assumption)

Reply:
Yes, not properly explained. We can accept a `Proposal` when it contains a `v.time` that, at the moment at which it was received, would not be considered `timely`. But we do that when the `Proposal` has a `POLRound > 0`, meaning that in a previous round (in the past), `v.time`was considered `timely` once received by a correct process.

In other words, `v.time` may not be `timely` now, as several communication steps passed since when it was originally proposed, but it was considered `timely` when it was originally proposed by at least a correct process, which is attested by the Proof-of-Lock associated to the received `Proposal`.